hedera-node/cli-clients/src/main/java/com/hedera/services/cli/signedstate/DumpAccountsSubcommand.java

```diff
@@ -0,0 +1,444 @@
+            final int limit,
+            @NonNull final Format format,
+            @NonNull final Verbosity verbosity) {
+        new DumpAccountsSubcommand(state, accountPath, limit, format, verbosity).doit();
+    }
+
+    @NonNull
+    final SignedStateHolder state;
+
+    @NonNull
+    final Path accountPath;
+
+    @NonNull
+    final Verbosity verbosity;
+
+    @NonNull
+    final Format format;
+
+    final int limit;
+
+    DumpAccountsSubcommand(
+            @NonNull final SignedStateHolder state,
+            @NonNull final Path accountPath,
+            final int limit,
+            @NonNull final Format format,
+            @NonNull final Verbosity verbosity) {
+        this.state = state;
+        this.accountPath = accountPath;
+        this.limit = limit >= 0 ? limit : Integer.MAX_VALUE;
+        this.format = format;
+        this.verbosity = verbosity;
+    }
+
+    void doit() {
+        final var accountsStore = state.getAccounts();
+        System.out.printf(
+                "=== %d accounts %s%n", accountsStore.size(), accountsStore.areOnDisk() ? "on disk" : "in memory");
+
+        final var accountsArr = gatherAccounts(accountsStore);
+
+        final var sb = new StringBuilder(1000);
+        final var reportSize = new int[1];
+
+        try (final var fileWriter = new FileWriter(accountPath.toFile(), StandardCharsets.UTF_8);
+                final var writer = new BufferedWriter(fileWriter)) {
+
+            if (format == Format.CSV) {
+                writer.write("account#");
+                writer.write(FIELD_SEPARATOR);
+                writer.write(formatCsvHeader(allFieldNamesInOrder()));
+                writer.newLine();
+            }
+
+            Arrays.stream(accountsArr).map(a -> formatAccount(sb, a)).forEachOrdered(s -> {
+                try {
+                    writer.write(s);
+                    writer.newLine();
+                } catch (final IOException ex) {
+                    System.err.printf("Error writing to '%s':%n", accountPath);
+                    throw new UncheckedIOException(ex);
+                }
+                reportSize[0] += s.length() + 1;
+            });
+        } catch (final IOException ex) {
+            System.err.printf("Error creating or closing '%s'%n", accountPath);
+            throw new UncheckedIOException(ex); // CLI program: Java will print the exception + stacktrace
+        }
+
+        System.out.printf("=== report is %d bytes%n", reportSize[0]);
+    }
+
+    @NonNull
+    HederaAccount[] gatherAccounts(@NonNull AccountStorageAdapter accountStore) {
+        final var accounts = new ConcurrentLinkedQueue<HederaAccount>();
+        final var processed = new AtomicInteger();
+        accountStore.forEach((entityNum, hederaAccount) -> {
+            final var n = processed.incrementAndGet();
+            if (n > limit) return;
+            accounts.add(hederaAccount);
+        });
+        final var accountsArr = accounts.toArray(new HederaAccount[0]);
```

Comment:
Interesting behavior here. I didn't know `ConcurrentLinkedQueue` would create a new array if the param passed in isn't big enough

Reply:
It's similar to how `List::toArray` works, yes.

Reply:
Apparently I need to read more javadoc